### PR TITLE
Implement vertical slice: Custom router

### PR DESCRIPTION
**Pull request by gunjpan**
This PR is a POC.

Connect to https://github.com/strongloop-internal/scrum-loopback/issues/714

### CODE REVIEW

File: lib/rest-adapter.js [hunk 1]
```diff
@@ -256,12 +257,27 @@ RestAdapter.prototype.createHandler = function() {
...
     debug('registering REST handler for class %j', className);
-
+    
+    // Mount the remoteClass router on all class routes.
+    restClass
+      .routes
+      .forEach(function (route) {
+        debug('    at %s', route.path);
+        root.use(route.path, router);
+        route.fullPath = restClass.getPath(); //add fullPath for Trie insertion
```
Comment (bajtos) [inline lib/rest-adapter.js]: This is smelly. Surely the full route of the path is not just the path of the class? Perhaps `route.classPath` would be a better property name?

File: lib/rest-adapter.js [hunk 2]
```diff
@@ -256,12 +257,27 @@ RestAdapter.prototype.createHandler = function() {
...
+        customRouter.registerPathAndHandlers(route);
+        //Do we need to register above class route?if yes, what's all method verb?
```
Comment (bajtos) [inline lib/rest-adapter.js]: > Do we need to register above class route?if yes, what's all method verb?

No, this does not make sense to me. There is no such thing as a class route. Only methods can contribute routes.

File: lib/rest-adapter.js [hunk 3]
```diff
@@ -286,15 +304,7 @@ RestAdapter.prototype.createHandler = function() {
...
-
-    // Mount the remoteClass router on all class routes.
-    restClass
-      .routes
-      .forEach(function(route) {
-        debug('    at %s', route.path);
-        root.use(route.path, router);
-      });
-
+
```
Comment (bajtos) [inline lib/rest-adapter.js]: Extra empty line, please remove.

File: lib/rest-router.js [hunk 4]
```diff
@@ -0,0 +1,36 @@
+module.exports = RestRouter;
+
+var Trie = require('./trie');
+
+function RestRouter(opts) {
+  this.Trie = new Trie();
```
Comment (bajtos) [inline lib/rest-router.js]: Variable and property names should start with a lower-case letter, i.e. `this.trie`.

File: lib/rest-router.js [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+  this.opts = opts;
```
Comment (bajtos) [inline lib/rest-router.js]: Don't abbreviate here, `this.options = options`.

File: lib/rest-router.js [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+    var methods = Trie.find(path);
+    if (methods && verb in methods) {
+      methods[verb](req, res, next);
+    }
+    else next();
+  }
+}
+
+function getPath(path) {
+  if (!this.opts || !(this.opts.caseSensitive)) {
```
Comment (bajtos) [inline lib/rest-router.js]: `this` is undefined in this function. I am proposing this:

```js
function normalizePath(path, options) {
  // your implementation, no change needed
};

// usage on L22:
var path = normalizePath(path, this.options);
```

File: lib/trie.js [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+      if (i + 1 === parts.length) {
+        node[parts[i]].methods = node[parts[i]].methods || {};
+        node[parts[i]].methods[verb] = handler;
+      }
+      node = node[parts[i]];
+    }
+  }
+  else { // if we would ever encounter this situation? 
+    node.methods = node.methods || {};
+    node.methods[verb] = handler;
```
Comment (bajtos) [inline lib/trie.js]: What happens when there is more than one handler registered for the same verb+path? Please check the current router does and implement the same behaviour. Don't forget to add unit-test.

File: lib/trie.js [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+
+//insert new nodes to Trie
+Trie.prototype.add = function (path, verb, handler) {
+  var node = this;
+  var parts = getPathParts(path);
+
+  if (parts.length) {
+    for (var i = 0; i < parts.length; i++) {
+      if (!node || !(parts[i] in node)) {
+        node[parts[i]] = new Trie();
```
Comment (bajtos) [inline lib/trie.js]: The first part of the condition will not work, as it distills to `if (!node) node[pars[i]] = new Trie()` - i.e. attempts to assign to an undefined variable.

File: lib/trie.js [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+      }
+      if (i + 1 === parts.length) {
+        node[parts[i]].methods = node[parts[i]].methods || {};
+        node[parts[i]].methods[verb] = handler;
```
Comment (bajtos) [inline lib/trie.js]: L18-19 and L25-26 are duplicated, please extract a shared private method.

An example of a possible API in action:

```js
node[parts[i]].addVerbHandler(verb, handler);
```

File: lib/trie.js [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+
+//Look for matching path and return handler methods for a match
+Trie.prototype.find = function (path) {
+  var node = this;
+  var parts = getPathParts(path);
+  for (var i = 0; i < parts.length; i++) {
+    if (!(parts[i] in node)) {
+      return false;
+    }
+    else {
```
Comment (bajtos) [inline lib/trie.js]: Coding style: merge L38 and L39 into a single line.

```js
} else {
```

File: lib/trie.js [hunk 11]
```diff
@@ -0,0 +1,57 @@
...
+      if (i + 1 === parts.length)
+        return node[parts[i]].methods;
+    }
+    node = node[parts[i]];
+  }
+}
+
+function getPathParts(path) {
+  path = path.trim();
```
Comment (bajtos) [inline lib/trie.js]: Is express router trimming whitespace too?

**Comment by bajtos** [comment]
The direction looks very reasonable to me. I am not going to review implementation details in respect to edge cases now, I need to see a test suite first.

File: lib/rest-adapter.js (hunk 2 above)
Comment (gunjpan) [inline lib/rest-adapter.js]: @bajtos: so, removing L277 would take care of the comments above

File: lib/trie.js (hunk 7 above)
Comment (gunjpan) [inline lib/trie.js]: @bajtos: I found that for a single `verb+path`, there is only one entry point in the express router. The router creates a stack of handler functions, which only invokes the handler, which was registered first by `router[method](path, handler)`, for the path. Other handlers won't see the light if they're not invoked if `request/response` cycle ends before `next()`.
So, for example, in the code below:
```
router.get('/planets', handlerOne());
router.get('/planets', handlerTwo());
```
only `handlerOne()` will be executed if it sends the `response` and doesn't reach the `next()`.
Thus, I believe, we will need to accommodate this case.
Thus, in the code above, at L19, an Array would be helpful to index registering of handlers for same verb+path:
`  node[parts[i]].methods[verb] = handler.push();`

Your thoughts please!

Comment (bajtos) [inline lib/trie.js]: Ah, of course, you are right. Your proposal to keep all handlers in an array makes perfect sense.

Comment (bajtos) [inline lib/trie.js]: Please make sure to add a unit-test for this case.

File: lib/rest-adapter.js (hunk 2 above)
Comment (bajtos) [inline lib/rest-adapter.js]: Makes sense.

File: lib/trie.js (hunk 11 above)
Comment (gunjpan) [inline lib/trie.js]: @bajtos: express router uses [`parseurl` module](https://github.com/pillarjs/parseurl/blob/master/index.js#L23) which uses a regular expression for verifying the url pattern (this pattern discards match for whitespace).

Using `parseurl` will be the right thing to do.

**Comment by gunjpan** [comment]
Closing this as this POC is used in Iteration-1 of custom rest-router implementation #282 
No further review is required for this PR (keeping the branch for now)
